Merchandise page

danielvancamp opened the pull request:
For : Add "Merch" Section to Static Page #6263

Q's 
- Best location to put css? 
- Added comment section how to add a seller, good practice or rather not

Info
- Just copy pasted some text (not the best writer in the world :P)
- Is responsive with bootstrap

danielvancamp committed:
added static merch page
added lang en .json
modified pages.js with the extra merch page

danielvancamp committed:
added static merch page
added lang en .json
modified pages.js with the extra merch page

danielvancamp committed:
added guide how to add merchandise

danielvancamp committed:
added guide how to add merchandise

danielvancamp committed:
added better bootstrap for multiple merchandises

danielvancamp committed:
added better bootstrap for multiple merchandises

danielvancamp committed:
updated counter to the real 2 sellers

danielvancamp committed:
added original static.styl again, mistakenly removed it while editing

danielvancamp committed:
Merge branch 'static-page-merch' of https://github.com/danielvancamp/habitrpg into static-page-merch

danielvancamp committed:
merge

danielvancamp committed:
once more i try to add it back

SabreCat commented on website/views/static/merch.jade:
```diff
@@ -0,0 +1,79 @@
+        .col-lg-6.col-md-6.col-sm-12
+            .merch-block
+                div.merch-header: img(src="/merch/"+env.t('merch-' + i + '-name')+"-logo.png")
+                div.img-con: a(href=env.t('merch-' + i + '-link') target='_blank'): img.img-rendering-auto.merch-img(src="/merch/" + env.t('merch-' + i + '-name') + ".png")
```
I'm only seeing broken image placeholders when testing this.

SabreCat commented on website/views/static/merch.jade:
```diff
@@ -0,0 +1,79 @@
+extends ./layout
+
+// Merchandise page : when adding a merch make sure to add following resources
+    + image of merchandise  --   env.t("merch-*-name").png       -- eg. teespring.png
```
Useful summary!

SabreCat commented on website/views/static/merch.jade:
```diff
@@ -0,0 +1,79 @@
+        - merch-*-goto     ->  Get myself a Habitica T-Shirts
+
+block vars
+    - var layoutEnv = env
```
We typically use 2-space indents throughout Habitica's code base, so tightening these up would help for consistency.

SabreCat commented:
We could use a link to this page in the footer. See https://github.com/HabitRPG/habitrpg/blob/develop/website/views/shared/footer.jade

crookedneighbor commented on website/views/static/merch.jade:
```diff
@@ -0,0 +1,79 @@
+
+
+block extraHead
+    style.
```
This should all be in the website/public/css

danielvancamp commented:
Thanks for feedback, i'll fix those things soon!
But first, food! ;)

danielvancamp committed:
updated tab indent
added link in footer
renamed Merchandise to Merch (looks better in the footer)
moved css to static.styl

danielvancamp commented:
Hey guys, 

Think I solved all the issues.
- Updated the indent/tab
- Moved css to website/public/css/static.styl
- Added link in the footer, at the bottom in the Company column --> Ok?
- Renamed "Merchandise" to "Merch" looked better in the footer --> Ok? 

I checked and though about why you can't see the images, but i got no clue. They should be in there. Still have that problem? I can see them without probs. 

Tell me if something isn't right yet, i'll fix it asap. 

Cheers!

Alys commented on common/locales/en/merch.json:
```diff
@@ -0,0 +1,14 @@
+{
+  "merch" : "Merch",
```
Could you please change the text on the right hand side to "Merchandise"?

Although probably wait until you hear more feedback so you can do it all at once. :)

Alys replied:
Although probably wait until you hear more feedback so you can do it all at once. :)

danielvancamp replied:
Noted! ;)

Alys commented:
@danielvancamp When we put this code on a test server that mimics production (not a local install), the images don't appear because there's their filenames are all lower-case while the code is specifying an initial upper-case letter. Looking through the code base, there's a wide range of styles for image file names, so it looks like we don't have any kind of standard. The best fix for this is to probably rename the image files, and also adjust the examples in the instructions, and probably add a specific note to the instructions about image file name case.

danielvancamp commented:
I see! Of course, seems logical. 

Maybe it's better to ask the image and the logo name in the json file? Or is that too much? 
So instead of making them with 
- ...(src="/merch/" + env.t('merch-' + i + '-name') + ".png)

Something like this
- ...(src="/merch/" + env.t('merch-' + i + '-goods))

and for the logo
- ...(src="/merch/" + env.t('merch-' + i + '-logo))

That way there is has more control over the naming but more to fill in as well. 

Other solution i think of is giving naming convention. For example
- always use capitals for each word
- when 2+ words, use spaces

eg. Teespring --> Teespring.png, Teespring-logo.png
eg. New Merch --> New Merch.png, New Merch-logo.png 

Sorry to bother you with the possible solutions, but i don't know the preferred ways just yet :) so better to ask i guess.

Alys commented:
Oh, I just realised another problem with putting the merchant names into the locales files, and with using them for creating image names - translators will try to translate them, and that will cause images to break, and also result in "translated" merchant names appearing on the page, which is undesirable. We need to come up with another way for specifying some of this config.

@crookedneighbor @lemoness Suggestions?

Alys commented:
Also, I think it's probably safest to manually specify the full file name for each image. That will give the most flexibility for putting as many images as needed on the page (there might be more than one per merchant). However the locale file is probably not the place for it. It might end up being necessary to create a static merchandise page, although I do like the way you made it dynamic with minimal data entry needed for each item.

danielvancamp commented:
Think you're probably right :) 
I'll wait a moment for @crookedneighbor and @lemoness reaction? Or shall i make it static already?

Alys commented:
Best to wait, and it might be a couple of days since they're in the USA and it's the thanksgiving weekend there.

danielvancamp commented:
No problem! 
One more thought! 

For the brand name of a merchant, those shouldn't get translated no? Like Treespring, will be the same name in every language. So maybe if we put a var list on the jade page with all the merchants in it. We can then use a sort of array to add the image names as well. 

For the multiple images for one merchant, isn't better to then just use the multiple items of that merchant and make one images with them (for example, if Treespring would make shoes as well, put a shoe in front of the t-shirt) and then link to all the merchandise of hibitica. 

Just shooting stuff to think about, no idea if they make some sense. I'll try to make it work in a var on the page already to see if that's something that makes more sense. Again even if it's shot down i won't mind. Fun to play around with :P

danielvancamp committed:
cleared merch.json of attributes that should not be translated

danielvancamp commented:
@Alys i think this could solve most of the issues we talked about. 
- image filenames are now  manually added
- no more string in the locale file that shouldn't be translated

What still needs to be tackled
- Multiple images (but as i asked before, maybe better to put multiple merchandise in one image?)

Don't feel obligated to answer right away! Just couldn't help trying something here :)!

danielvancamp committed:
minor update in the comment section

Alys commented:
@danielvancamp I like how you've done this, and I think the one-image idea is potentially good, although I'm not the person who'd be adding the data and creating the image, so my opinion counts for little. ;)

@lemoness and @SabreCat :

The proposal is that whenever a new merchant is added, you would do these three things:

(1) Edit the "merch.json" locale file to provide two pieces of text for translation. For example:

```
"merch-0-summary" : "Teespring is a platform that makes it easy for anyone to create and sell high-quality products people love, with no cost or risk.",
"merch-0-goto" : "Get myself a Habitica T-Shirts",
```

(2) Edit the jade script `website/views/static/merch.jade` to insert one line for the merchant, to specify information that must not be translated (merchant name, URL, logo file location, and merchandise image location):

```
merchants.push({name: 'Teespring', link: 'https://teespring.com/habitica-gryphon-t-shirt', image: 'teespring.png', logo: 'teespring-logo.png'})
```

(3) Create just **one** image to display the items available from that merchant and upload it to the `website/public/merch/` directory. If you wanted to display two or more items, you would need to put them all into a single image.

This is just the current proposal, so please tell us if you don't think it will work for you. Once we have your approval or otherwise for this, I'll do another test of the code and upload a screenshot for your comments about how it looks.

crookedneighbor commented on common/locales/en/merch.json:
```diff
@@ -0,0 +1,10 @@
+{
+  "merch" : "Merchandise",
+
+  "merch-0-summary" : "Teespring is a platform that makes it easy for anyone to create and sell high-quality products people love, with no cost or risk.",
```
I did something like this for the FAQ strings and I am now regretting it. It locks you into a particular order, and it makes it hard to retire a piece of merchandise. 

I propose something like: 

`merch-teespring-summary`

crookedneighbor replied:
Then in the jade file, you'd have something like:

``` jade
- var merchants = []
- merchants.push({key: 'teespring', url: 'https://teespring.com/habitica-gryphon-t-shirt'})

each merchant, index in merchants
  if (index + 1) % 2 !== 0
    .row

  .col-md-6.col-sm-12
    .merch-block
      .merch-header: img(src="/merch/#{merchant.key}-logo.png")
      .merch-img-con: a(href=merchant.url target='_blank'): img.img-rendering-auto.merch-img(src="/merch/#{merchant.key}-sample.png")
      p=env.t('merch-#{merchant.key}-summary')
      a.btn.btn-habitica.merch-link(href=merchant.url target='_blank')=env.t('merch-#{merchant.key}-goto')
```

Then, if later on we wanted to add a separate product from teespring, we could just add a new key, `teespring2` and add

``` jade
- merchants.push({key: 'teespring2', url: 'https://teespring.com/other-url'})
```

And the corresponding translation strings.

danielvancamp replied:
Hey @crookedneighbor 

Just updated this part, just checking something first i discussed with @Alys. 
I can make the links to the logo and the merchandise image dynamic like you suggested, but that means : when a image is added, the name of the image has to be exactly typed a certain way (eg. teespring,png). Is this something desirable? The way it is now, they just type over the image name and can verify the name is correct themselves. 
I can find myself in both options, when clear guidelines are presented i think the dynamic should be no problem. 

Just let me know and i'll put in what you guys think is best! 

Cheers (and a little to late but happy Thanksgiving! ;))

crookedneighbor commented on website/public/css/static.styl:
```diff
@@ -69,3 +69,39 @@ a.h2.accordion.collapsed:before
+    max-width: 350px
+    padding: 30px 0
+
+.merch-link
```
I don't think these extra classes are necessary. They're all nested under `.merch-block`, so you could do:

``` stylus
.merch-block
  border: 1px dotted #553889
  border-radius: 6px
  padding: 15px
  margin-bottom: 20px

  .btn
    margin-top: 20px
    background-color: #553889
    color: #ffffff

    &:hover
      color: #ffffff
      background-color: #452d6e
      border-color: #342253
```

crookedneighbor commented on website/public/css/static.styl:
```diff
@@ -69,3 +69,39 @@ a.h2.accordion.collapsed:before
+
+.merch-img-con
+  text-align: center
+  .merch-img
```
Why not just?

``` stylus

.merch-img-con
  text-align: center

  img 
    max-width: 350px
    padding: 30px 0    

```

danielvancamp commented:
Nice! 
I'll fix those thing once i get home tonight.
Thanks for the feedback!

SabreCat commented:
Yay, recent changes have the images showing up when run on test servers!

I'd suggest these simplifications and corrections for the button captions ( @lemoness may have further thoughts):
"Get myself a Habitica T-shirts" => "Get a Habitica T-shirt"
"Get myself a Habitica sticker" => "Get Habitica stickers"

SabreCat commented:
The button and image for Stickermule link to the Teespring page.

Alys commented:
@danielvancamp Lemoness has approved the current situation for how the information is added (summarised in https://github.com/HabitRPG/habitrpg/pull/6307#issuecomment-160515959) so we'll go with that. Thank you! It is a good idea!

danielvancamp committed:
updated static.styl with less classes
added key attribute to merchant array
changed item to merchant for clearity
added fallback for when no logo is found (no image can be provided)
updated guidelines
updated json

danielvancamp commented:
Hey everyone! 

@Alys oow yeah! 

I've made most of the changes already. Here a summary : 
- updated static.styl with less classes, like @crookedneighbor recommended
- added key attribute to merchant array, like @crookedneighbor also recommended
- changed item to merchant for clarity, like @crookedneighbor showed me the way :D
- added fallback for when no logo is found (no image can be provided)
- updated guidelines
- updated json
- made the button 100% with for more alignment (tell me if you want it the other way again)

What i didn't change yet but can easily be fixed
- changed the way the merchandise image and merchant logo is loaded

This is still with manual input from the user, i can change it to the dynamic input like this : 
- merchant.key = teespring  -->  "#{merchant.key}" + "(-logo.)png"  -->  teespring.png, teespring-logo.png 

I'll hear it when you guys know! 
Again thanks for feedback and hope not to absorb to much from your time with this page. 
Cheers!

SabreCat commented:
Given that the Teespring campaign is about to end, we may as well remove its entry from the page, leaving the Stickermule bit. Sorry the review process has taken so long!

danielvancamp commented:
I'll remove it in the next pull! 
No problem, fully understandable :)
Once i got more insight on this comment https://github.com/HabitRPG/habitrpg/pull/6307#issuecomment-160769381 i'll push the updates! 

Cheers! and (for me at least) goodmorning!

crookedneighbor commented:
I think it's fine to require the image files to have a particular filename. 

Is that the only point you needed feedback on?

danielvancamp commented:
Jep! 
I'll change it tomorrow like that and put an update online! 
One more Q, the Teespring campaign is offline but you can still show your interest, shall i still remove it or leave it on there? 
Thanks for the feedback! 

Goodnight

danielvancamp committed:
Merge branch 'develop' into static-page-merch

danielvancamp committed:
update logo and image links to be dynamicly added instead with a manual written link

danielvancamp commented:
Hey guys! 
Updated the page like @crookedneighbor asked! 
So now the images and logos are added like this 
- MERCHANT.KEY.png
- MERCHANT.KEY-logo.png 
- MERCHANT.KEY = teespring
- teespring.png
- teespring-logo.png

This states that we require the filenames to be like this exact! Always png's and always in that order. 

Something else : @SabreCat I didn't remove the teespring campaign just yet until i hear you guys about it. You can still backorder it at their site, so leave it on or rather take it out?

One more open question, i just merged with the latest upstream:develop branch, is that something that needs to be done or not really? Just to know how to do this in the future! 

Hear you guys later!
Cya

lemoness commented:
Do we know how the back orders work? If they require us to launch another
campaign, they might not be worth it. Otherwise, I am fine with that.

On Monday, December 14, 2015, Daniel Van Camp notifications@github.com
wrote:

> Hey guys!
> Updated the page like @crookedneighbor
> https://github.com/crookedneighbor asked!
> So now the images and logos are added like this
> - ## MERCHANT.KEY.png
>   
>   ## MERCHANT.KEY-logo.png
>   
>   ## MERCHANT.KEY = teespring
>   
>   teespring.png
> - teespring-logo.png
> 
> This states that we require the filenames to be like this exact! Always
> png's and always in that order.
> 
> Something else : @SabreCat https://github.com/SabreCat I didn't remove
> the teespring campaign just yet until i hear you guys about it. You can
> still backorder it at their site, so leave it on or rather take it out?
> 
> One more open question, i just merged with the latest upstream:develop
> branch, is that something that needs to be done or not really? Just to know
> how to do this in the future!
> 
> Hear you guys later!
> Cya
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/6307#issuecomment-164556127.

danielvancamp commented:
Hey @lemoness 

I just checked with the guys from teespring, should be as follows : 

_Once a campaigns ends, if the designer has chosen the auto-relaunch option, the button that used to say “Campaign Ended” will instead say “Wait! I Still Want This!”
This process holds on to (or reserves) items until a minimum of 10 reservations have been placed. At that point in time, the campaign will “auto-relaunch” – becoming active again.
When this happens, the campaign will re-activate and run for 10 more days. That way, even more people have the opportunity to order if they missed it the first time.
Along the way, you’ll still receive all the normal confirmation emails that you typically get with Teespring (confirmation of the order, confirmation that the campaign reached it’s goal, your transaction status, and confirmation that the order has shipped.)
For more detailed information, you can take a look at our Teespring Answers page: http://answers.teespring.com/customer/portal/articles/1710931-relaunched-campaigns_

Do you guys know there is one other campaign active? (or is it someone else?)
- https://teespring.com/habitica-gryphon-hoodie-a-2404#pid=212&cid=5819&sid=front

Those are the ones that were closed already : 
- https://teespring.com/habitica-gryphon-onesie#pid=287&cid=6508&sid=front
- https://teespring.com/habitica-gryphon-t-shirt_eu
- https://teespring.com/habitica-gryphon-t-shirt

If you let me know what you guys want to do with this, i'll update the page. I can put all the campaigns in with ease! 

Cheers! Happy holidays!

lemoness commented:
Thanks so much for looking into this!

The Gryphon Hoodie is NOT our campaign, which is pretty gross, because it
means someone is trying to profit from our logo. We're looking into that
right now. We really appreciate you letting us know about it!

As for the others, go ahead and leave the links on the merch page! We will
just let them refresh periodically.

On Thu, Dec 24, 2015 at 1:25 AM, Daniel Van Camp notifications@github.com
wrote:

> Hey @lemoness https://github.com/lemoness
> 
> I just checked with the guys from teespring, should be as follows :
> 
> _Once a campaigns ends, if the designer has chosen the auto-relaunch
> option, the button that used to say “Campaign Ended” will instead say
> “Wait! I Still Want This!” This process holds on to (or reserves) items
> until a minimum of 10 reservations have been placed. At that point in time,
> the campaign will “auto-relaunch” – becoming active again. When this
> happens, the campaign will re-activate and run for 10 more days. That way,
> even more people have the opportunity to order if they missed it the first
> time. Along the way, you’ll still receive all the normal confirmation
> emails that you typically get with Teespring (confirmation of the order,
> confirmation that the campaign reached it’s goal, your transaction status,
> and confirmation that the order has shipped.) For more detailed
> information, you can take a look at our Teespring Answers page:
> http://answers.teespring.com/customer/portal/articles/1710931-relaunched-campaigns
> http://answers.teespring.com/customer/portal/articles/1710931-relaunched-campaigns_
> 
> Do you guys know there is one other campaign active? (or is it someone
> else?)
> 
>    -
>    https://teespring.com/habitica-gryphon-hoodie-a-2404#pid=212&cid=5819&sid=front
> 
> Those are the ones that were closed already :
> 
>    -
>    https://teespring.com/habitica-gryphon-onesie#pid=287&cid=6508&sid=front
> - https://teespring.com/habitica-gryphon-t-shirt_eu
> - https://teespring.com/habitica-gryphon-t-shirt
> 
> If you let me know what you guys want to do with this, i'll update the
> page. I can put all the campaigns in with ease!
> 
> Cheers! Happy holidays!
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/HabitRPG/habitrpg/pull/6307#issuecomment-167076510.

danielvancamp commented:
No problem at all! 
I'll resolve the conflicts tomorrow and then i think it it finished, unless i missed something. 

Hope you get that bad campaign straightened out! 

Cheers!

danielvancamp committed:
Merge branch 'develop' into static-page-merch

danielvancamp committed:
added EU version of t-shirt

danielvancamp commented:
Hey guys! 

Sorry for the late action, had some major unforeseen work at the office. Bu i had it on my todo (habitica tasks) and can check it off right now!

I left out the onsie at this moment because it didn't have a picture anymore in the campaign, so i couldn't make a image for it. 

All other things should be as intended. Make sure the check the texts underneath the images. 

Cheer!

SabreCat commented:
Merged at last! I have awarded you your second contributor tier, given that this took quite a bit of back-and-forth effort. Thanks much!

danielvancamp commented:
Very glad to help! 
And thanks a lot for the new tier. On to the next one!
